NeoML/src/Dnn/Layers/LowRankAdaptedFullyConnectedLayer.cpp

```diff
@@ -0,0 +1,311 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#include <common.h>
+#pragma hdrstop
+
+#include <NeoML/Dnn/Layers/LowRankAdaptedFullyConnectedLayer.h>
+#include <NeoML/Dnn/Layers/ActivationLayers.h>
+#include <NeoML/Dnn/Layers/DropoutLayer.h>
+#include <NeoML/Dnn/Layers/EltwiseLayer.h>
+#include <NeoMathEngine/NeoMathEngine.h>
+
+namespace NeoML {
+
+static const char* const baseFcName = "FullyConnectedBase";
+static const char* const loraFcAName = "FullyConnectedALoRA";
+static const char* const loraFcBName = "FullyConnectedBLoRA";
+static const char* const loraDropoutName = "DropoutLoRA";
+static const char* const loraScalingName = "ScalingLoRA";
+static const char* const loraSumName = "SumLoRA";
+
+//----------------------------------------------------------------------------------------------
+
+CLowRankAdaptedFullyConnectedLayer::CLowRankAdaptedFullyConnectedLayer( IMathEngine& mathEngine, const char* name ) :
+	CCompositeLayer( mathEngine, ( name == nullptr ) ? "CLowRankAdaptedFullyConnectedLayer" : name )
+{
+	buildLayer();
+}
+
+CLowRankAdaptedFullyConnectedLayer::CLowRankAdaptedFullyConnectedLayer( CFullyConnectedLayer& fc ) :
+	CCompositeLayer( fc.MathEngine(), fc.GetName() ),
+	baseFc( &fc )
+{
+	baseFc->SetName( baseFcName );
+	SetInputMapping( *baseFc );
+	SetOutputMapping( *baseFc );
+	AddLayer( *baseFc );
+
+	NeoPresume( baseFc != nullptr && baseFc->GetName() == CString( baseFcName ) );
+}
+
+static const int LowRankAdaptedFullyConnectedLayerVersion = 0;
+
+void CLowRankAdaptedFullyConnectedLayer::Serialize( CArchive& archive )
+{
+	archive.SerializeVersion( LowRankAdaptedFullyConnectedLayerVersion );
+	CCompositeLayer::Serialize( archive );
+
+	archive.Serialize( loraRank );
+	archive.Serialize( loraAlpha );
+	archive.Serialize( loraDropoutRate );
+
+	if( archive.IsLoading() ) {
+		baseFc = CheckCast<CFullyConnectedLayer>( GetLayer( baseFcName ) );
+		// optional LoRA layers
+		loraFcA = HasLayer( loraFcAName )
+			? CheckCast<CFullyConnectedLayer>( GetLayer( loraFcAName ) )
+			: nullptr;
+
+		if( loraFcA != nullptr ) {
+			loraFcB = CheckCast<CFullyConnectedLayer>( GetLayer( loraFcBName ) );
+			loraDropout = HasLayer( loraDropoutName )
+				? CheckCast<CDropoutLayer>( GetLayer( loraDropoutName ) )
+				: nullptr;
+			loraScaling = CheckCast<CLinearLayer>( GetLayer( loraScalingName ) );
+			loraSum = CheckCast<CEltwiseSumLayer>( GetLayer( loraSumName ) );
+		} else {
+			NeoAssert( loraRank == 0 && loraAlpha == 0 && loraDropoutRate == 0 );
```

Comment:
Yes, it may be valid, but not-reshaped layers throw an exception on their serialization.
If the `BuildLora()` is called, so `loraFcA != nullptr` and `loraRank != 0`, else `loraFcA == nullptr` and `loraRank == 0`. 
I do not know, how to check: either the layer be serialized (was reshaped) now, or not.

Reply:
fixed